graph/propertyconventions.md

```diff
@@ -0,0 +1,70 @@
+# Guidelines for the conventions of properties on graph types
+
+## Updating read-only properties
+
+### Convention
+
+It is a common call pattern for Microsoft Graph clients to `GET` an entity, make changes to a few properties, and then `PATCH` the same entity with the updated values.
+When clients do this, they often ignore that some properties are read-only or that a property may have a value that is allowed in a response, but not in a request.
+Because of this, workloads should support the practice that a request that updates a property should be considered a no-op if the value in the request matches the current value of the property.
+
+### Example
+
+For the following schema:
+```xml
+<EntityType Name="entity">
+  <Key>
+    <PropertyRef Name="id" />
+  </Key>
+  <Property Name="id" Type="Edm.String" Nullable="false" />
+  <Property Name="foo" Type="Edm.String" Nullable="false" /> <!--this property is read-only-->
```

Comment:
Q: are there plans to add a `ReadOnly="true"` to the `Property` Node where applicable (or else how can clients know this is a readonly field)?

Reply:
Hello! There are three ways to indicate this in the CSDL:
1. The `Computed` [annotation](https://github.com/oasis-tcs/odata-vocabularies/blob/main/vocabularies/Org.OData.Core.V1.xml#L384)
2. The `Immutable` [annotation](https://github.com/oasis-tcs/odata-vocabularies/blob/main/vocabularies/Org.OData.Core.V1.xml#L380)
3. The `NonUpdatableProperties` [property](https://github.com/oasis-tcs/odata-vocabularies/blob/1fd9313b8c747db3707562f3a56c955aaf15434f/vocabularies/Org.OData.Capabilities.V1.xml#L713) of `UpdateRestrictionsType`

It's also worth noting what the [standard](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#_Toc31358979) says about this:
> Key and other properties marked as read-only in metadata [...] can be omitted from the request. If the request contains a value for one of these properties, the service MUST ignore that value when applying the update.

Our behavior in graph doesn't actually match any of these. In graph, if a read-only property is provided, we check that it has the same value as the current property; if it does, we ignore it, and if it doesn't we produce an error. As a result of investigating all of this (thanks for making me look into it!), we have created [this](https://issues.oasis-open.org/browse/ODATA-1543) issue with the OASIS technical committee to discuss how to represent this behavior in OData. Most likely, we will update the `Immutable` annotation to be more flexible, changing it from always ignoring the property when provided to allowing it to be validated if provided.